base/compiler/abstractinterpretation.jl

```diff
@@ -606,6 +606,9 @@ function abstract_call_method(interp::AbstractInterpreter, method::Method, @nosp
     if edge === nothing
         edgecycle = edgelimited = true
     end
+    edge_effects = override_edge_effects(edge_effects, method, :consistent)
+    edge_effects = override_edge_effects(edge_effects, method, sv, :effect_free)
```

Comment:
Actually, upon second thought, I think the same argument as to `:consistent`, applies to `:effect_free` as well, so I think these should all just look at the callee's effects. However, in that case, I'm not completely sure why we need this at all, because the callee's effects should have already caused an override at the conclusion of inference.

Reply:
> However, in that case, I'm not completely sure why we need this at all, because the callee's effects should have already caused an override at the conclusion of inference.

That is because we don't respect annotated effect settings when we update effect property. So even if a callee frame is explicitly annotated as `:effect_free`, out effect tracking may taint its `:effect_free` property. This commit basically just overrides that entire result here.

And now I think we should move this mechanism to `tristate_merge!`?

Reply:
Well, I thought we were doing this overriding just after inference is done, but before writing the effects into the CodeInstance for caching? I guess that code got lost somewhere? In any case, that's the correct place to put this. The termination override here is just because otherwise you may have to taint something due to recursion, which the termination override prevents.

Reply:
Makes sense. I will (re-)implement the override mechanism just before `CodeInstance` caching.